### PR TITLE
Enh/c+r

### PR DESCRIPTION
**Pull request by NiranjanChaudhari0929**
Creating a "Draft Pull Request" so that you can review the code and let me know for any updates.

### CODE REVIEW

**Review by ajtritt — CHANGES_REQUESTED** [review]

File: scripts/.ipynb_checkpoints/train_simple-checkpoint.sh
[the remark this answers is not in the document]
Comment (NiranjanChaudhari0929) [reply scripts/.ipynb_checkpoints/train_simple-checkpoint.sh:1]: I will delete this

File: src/lmcontrol/data_utils.py [hunk 1]
```diff
@@ -6,58 +8,65 @@
...
         logger.debug(f"Found the following keys in {npz_path}: {' '.join(sorted(md_keys))}")
-        for k in md_keys:
-            v = metadata.setdefault(k, list())
-            v.extend([str(npz[k])] * len(npz['masks']))
+        
+        for k in sorted(md_keys):
+            if npz[k].ndim == 0:
+                metadata.setdefault(k, []).extend([str(npz[k])] * len(indices))
+            else:
+                metadata.setdefault(k, []).extend(np.array(npz[k])[indices])
```
Comment (ajtritt) [inline src/lmcontrol/data_utils.py:42]:
```suggestion
        for k in md_keys:
            v = metadata.setdefault(k, list())
            v.extend([str(npz[k])] * len(npz['masks']))
```

Comment (ajtritt) [reply src/lmcontrol/data_utils.py:42]: check to make sure this works.

**Review by ajtritt — CHANGES_REQUESTED** [review]

File: src/lmcontrol/nn/clf.py [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+            'time': nn.MSELoss(),
+            'sample': nn.CrossEntropyLoss(),
+            'condition': nn.CrossEntropyLoss(),
+            'feed': nn.CrossEntropyLoss(),  
+            'starting media': nn.CrossEntropyLoss()  
+        }
+        if label_type not in self.loss_functions:
+            raise ValueError(f"Unknown label type: {label_type}. Expected one of {list(self.loss_functions.keys())}")
+        
+        self.backbone = resnet(weights=None, progress=True, label_type=label_type, block=block, layers=layers, planes=planes, num_outputs=num_outputs, return_embeddings=return_embeddings)
```
Comment (ajtritt) [inline src/lmcontrol/nn/clf.py:81]:
```suggestion
        self.backbone = ResNet(label_type=label_type, block=block, layers=layers, planes=planes, num_outputs=num_outputs, return_embeddings=return_embeddings)```

File: src/lmcontrol/nn/clf.py [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+        **kwargs: parameters passed to the ``torchvision.models.resnet.ResNet``
+            base class. Please refer to the `source code
+            <https://github.com/pytorch/vision/blob/main/torchvision/models/resnet.py>`_
+            for more details about this class.
+
+    .. autoclass:: torchvision.models.ResNet18_Weights
+        :members:
+    """
+    return _resnet(label_type=label_type, block=block, layers=layers, planes=planes, num_outputs=num_outputs, weights=weights, progress=progress, return_embeddings=return_embeddings)
+
```
Comment (ajtritt) [inline src/lmcontrol/nn/clf.py:58]:
```suggestion
```

File: src/lmcontrol/nn/clf.py [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+from torch.utils.data import DataLoader
+from lightning.pytorch.callbacks import ModelCheckpoint
+from torchvision.models.resnet import BasicBlock, Bottleneck 
+from typing import Type, Union, List, Optional, Callable, Any
+from optuna.integration import PyTorchLightningPruningCallback
+
+from ..utils import get_logger
+from ..data_utils import load_npzs, encode_labels
+from .dataset import LMDataset, get_transforms as _get_transforms
+from lmcontrol.nn.resnet import _resnet
```
Comment (ajtritt) [inline src/lmcontrol/nn/clf.py:36]:
```suggestion
from lmcontrol.nn.resnet import ResNet
```

File: src/lmcontrol/nn/dataset.py [hunk 5]
```diff
@@ -70,34 +75,70 @@ def __init__(self, npzs, use_masks=False, return_labels=False, logger=None):
...
+            self.data = self.data[val_indices]
+            if self.labels is not None:
+                self.labels = self.labels[val_indices]
 
     def __getitem__(self, i):
         ret = self.data[i]
         if self.transform is not None:
             ret = self.transform(ret)
         labels = -1 if self.labels is None else self.labels[i]
+
```
Comment (ajtritt) [inline src/lmcontrol/nn/dataset.py:141]:
```suggestion
```

File: src/lmcontrol/nn/resnet.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    planes: List[int],
+    num_outputs: int,
+    return_embeddings: bool = False,
+    **kwargs: Any,
+) -> ResNet:
+        
+    model = ResNet(label_type=label_type, block=block, layers=layers, planes=planes, num_outputs=num_outputs, return_embeddings=return_embeddings)
+
+
+    return model
```
Comment (ajtritt) [inline src/lmcontrol/nn/resnet.py:182]:
```suggestion
```

**Review by ajtritt — CHANGES_REQUESTED** [review]

**Review by ajtritt — APPROVED** [review]

**Pull request merged**